### ISSUE TITLE
(microservices) configurable message parsing?

### ISSUE THREAD

**Issue by eckdanny** [nestjs/nest#4407]
## Feature Request

Is it feasible to make message parsing/formatting in nest microservices configurable?

I'm working with `transport.RMQ` and I hit a wall because my message Producers (another system) publish `content_type: text/xml` -formatted messages.

I don't want to conflate concepts (and maybe I'm wrong here) but AFAIK:

| concept | description |
| :--- | :---| 
| message parsing/formatting | _i.e.;_ do to/from [`msgPack`](https://msgpack.org/), `xml`, `json` operations |
| serialization/deserialization | _i.e.;_ _"new-up"_ or _to-plain-obj_ db entities |

Can there be a configuration point for message formatting/parsing? The hook I think I'm looking for is like in [`server-mqtt#128`](https://github.com/nestjs/nest/blob/6b3c932193512ad56a752ac87ad72d2bf8f3a0df/packages/microservices/server/server-mqtt.ts#L128) but perhaps could be defined in `createMicroservice` like:

```ts
NestFactory.createMicroservice(
   AppModule,
   {
      transport: Transport.RMQ,
      options: {
         ...restOptions,
         parse: (msg: ConsumeMessage) => parseMessageSomehow(msg),
      }
   }
)
```

In my particular case (consuming XML-formatted msgs) my microservice barfs [here](https://github.com/nestjs/nest/blob/6b3c932193512ad56a752ac87ad72d2bf8f3a0df/packages/microservices/server/server-rmq.ts#L110). In the general use-case, I think a `parse/format` message "hook" could be really handy as an interceptor to adapt nest microservices to fit in with an existing message-based ecosystem.

Did i miss a part in the docs how to accomplish this or does the feature not exist? Thanks!

**Comment by eckdanny** [comment]
Encountered another use case that falls into this bucket just now:

_Some_ producers out there (e.g.; https://github.com/Particular/NServiceBus.Newtonsoft.Json/issues/53) include BOM in the serialized JSON msg body. 🤔

A fix on the consumer side with this proposal (reusing the block from ^) would be simple:

```js
const stripBom = require('strip-bom');

const parseMessageSomehow = (msg) => ({
   ...msg,
   content: JSON.parse(stripBom(msg.content.toString())),
}
```

**Comment by eckdanny** [comment]
@johnbiundo just saw [your comment in 406](https://github.com/nestjs/docs.nestjs.com/issues/406#issuecomment-593559176) and skimmed your [dev.to article "Techniques for Application Integration with NestJS..."](https://dev.to/nestjs/integrate-nestjs-with-external-services-using-microservice-transporters-part-3-4m20). _Super_ helpful + relevant to what I'm trying to do...

Thoughts on discriminating message `parse/format` and `serialize/deserialize` operations? I was thinking the "hook" (adaptor) to convert msg formats in/out of the "nest" format mentioned in your article is more of the former. 

Maybe the NATS server is implemented differently, but I can't hook into deserializers in the nest server-rmq impl anyway b/c it hits _after_ parsing ([`server-rmq#110@7.0.5`](https://github.com/nestjs/nest/blob/6b3c932193512ad56a752ac87ad72d2bf8f3a0df/packages/microservices/server/server-rmq.ts#L110-L111))

**Comment by johnbiundo** [comment]
@eckdanny without digesting this whole thread, perhaps what you need to do is build a custom transporter that simply overrides `ServerRMQ#handleMessage`.  This might be a better article series for you, maybe starting here: https://dev.to/nestjs/part-2-basic-server-component-onk#using-a-custom-transporter-strategy

**Comment by johnbiundo** [comment]
On a quick read, your thought process (on creating a hook for parsing) makes sense to me too, but I think brute forcing it with a custom transporter might be your best bet in the short term.

**Comment by johnbiundo** [comment]
Actually, if @kamilmysliwiec agrees with this, it should be pretty easy to put such a hook in.  Is that something you'd be willing to build?

**Comment by shlomiassaf** [comment]
I'm not sure a hook is the right solution, given the current implementation.

I think that the assumption that all microservices are request/response oriented is wrong since there are a lot of flows which might not follow this approach.

The serializer & deserializer should not confine to the `Consumer` model, they should play a more generic role, see #4406

The implementation is a bit wrong here:

https://github.com/nestjs/nest/blob/6b3c932193512ad56a752ac87ad72d2bf8f3a0df/packages/microservices/server/server-rmq.ts#L110-L111

The call to `JSON.parse` should be done in the deserializer, after all `JSON.parse` is a **deserializer**.

The problem here is that the `serializer` & `deserializer` are only ment to go to/from the `Consumer` model.

This limits the possibilities of working with microservices only to that model.
 
I'm not sure what's the best design here, maybe for example, having an rmq server class and another rmq-consumer server class which takes the consumer model... maybe not..

**Comment by johnbiundo** [comment]
@shlomiassaf I haven't yet tried to read through your other issue, but could you please define the term *consumer model* that you're using?  I haven't worked much with RMQ if it's a term that is common in that context.  

I've generally been using the terms *requestor* and *responder* to describe the *roles* (within the existing capabilities of Nest microservices) in an exchange between two parties.  E.g., a requestor issues requests/emits events through a `ClientProxy` and a responder listens for incoming events/requests in an instance of `NestMicroservice`.  

It sounds like you're alluding to a more general set of communication patterns than those currently supported, but I think I need to better understand your terminology to understand what you're proposing.

**Comment by shlomiassaf** [comment]
@johnbiundo Yes, I refer to the terms  requestor and responder. (you can read the issue I posted, its not long).

Right now the entire microservices module is focused around that model (+ Events...)

Since `Server` can be used for other paradigms I think it should be separated from that model.

**Comment by johnbiundo** [comment]
@shlomiassaf OK, read your other issue.  I think I get it, but could you perhaps make this more concrete by providing an example use case for some other paradigm you'd envision supporting?

**Comment by shlomiassaf** [comment]
i'm currently building a wrapper for Azure ServiceBus which is a bit different.
It requires pre existing topics/queues.
It has built it rules, etc...
It requires management integration.
It has some integrated flows (i.e. PeekLock vs FetchAndDelete read mode)

While using the  requestor / responder model does exist within the message model the message handling process (complete/abandon/deal letter/defer) is different.

There is also a built in retry process and an **autoComplete** feature which **must** be handled if it's false.

On top of the basic implementation of allowing decorators to mark events handlers for queues/topics I also want to allow additional control over the flow.

For example, adding a backoff mechanism...

**In other words, I would like to make the implementation itself super simple and let the user control the behavior**

What I have in mind is to use that implementation along with `@nestjs/cqrs` to create sagas that will manage the behavior.

---

Going back to my suggestion, if the implementation is simple we can build sages to do virtually anything. E.G build a saga for each server (rmq, nats, kafka...) and use that instead of hardcoding that specific behavior to the server.

In any case, please do not add more design specific code into the server and let it be generic as possible so we can implement for different scenarios.

**Comment by johnbiundo** [comment]
@shlomiassaf Thanks for the details.  To be clear, I have an interest in how the microservice subsystem is architected, but it's entirely up to @kamilmysliwiec to decide how he'd like to consider this input.

**Comment by shlomiassaf** [comment]
I'm sure people are invested in this, no doubt.

I'm not even remotely suggest to change the architecture from the ground up, i'm just suggest to provide another level so to `Server` is not tied up to a single paradigm.

Think about it like this.

We strip `Server` out of all of the `MessagPattern` implementation and leave only `EventPattern`.
This will remove serializer, deserializer, send and more...
It will now act as a registry for route handling of the events registered.

This will be a simple class which will allow anyway to write a custom transporter that handles events and whatever he wants.

Now, we create a new class, `ServerRequestResponse` which will inherit from `Server` and will implement all the thing currently in `Server`... 

Change all transporters to inherit from `ServerRequestResponse` and we're done.

**Comment by kamilmysliwiec** [comment]
I think we've got a little lost in the main subject.

@eckdanny as @johnbiundo suggested, it should be fairly simple to move this line https://github.com/nestjs/nest/blob/6b3c932193512ad56a752ac87ad72d2bf8f3a0df/packages/microservices/server/server-rmq.ts#L110-L111 into the `deserialize()` method. Then, you should be able to override the parsing logic on your side. Would you like to create a PR for this issue?

**Comment by eckdanny** [comment]
Sure. I haven't contrib'd before but the code change sounds easy enough. Should have some time one of these next few evenings 👍

**Comment by kamilmysliwiec** [comment]
Great! Let me know if you have any questions @eckdanny

**Comment by kamilmysliwiec** [comment]
Let's track this here https://github.com/nestjs/nest/pull/4465
